### PR TITLE
proj: 8.2.1 -> 9.0.0

### PR DESCRIPTION
**Pull request by willcohen**
###### Description of changes

Updates `proj` to 9.0.0. Creates a new `proj_8` for the existing derivation, and updates its `hash` attribute to `sha256`, noticed when trying to use `lib.fakeSha256` during the update.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by willcohen** [comment]
My local build on my darwin machine is failing with the following sorts of errors:
```
 1/57 Test  #1: geodesic-test ....................Subprocess aborted***Exception:   0.11 sec
dyld[49253]: Library not loaded: @rpath/libproj.25.dylib
  Referenced from: /private/tmp/nix-build-proj-9.0.0.drv-0/source/build/bin/geodtest
  Reason: tried: '/nix/store/pll7kallqnbxhpjb6v0ql0kjndibwiyz-swift-corefoundation-unstable-2018-09-14/Library/Frameworks/libproj.25.dylib' (no such file), '/nix/store/pll7kallqnbxhpjb6v0ql0kjndibwiyz-swift-corefoundation-unstable-2018-09-14/Library/Frameworks/libproj.25.dylib' (no such file), '/usr/local/lib/libproj.25.dylib' (no such file), '/usr/lib/libproj.25.dylib' (no such file)
```

I see no reason why the 9.0 build should be causing this where 8.2 didn't so I'd like to see if this is just a local darwin issue to my machine (Monterey, stars aligning, etc) and whether ofborg is able to have it build similarly to 8.2.

**Comment by risicle** [comment]
Is there actually anything using `proj_8`? We usually only keep multiple versions of something if it has an in-nixpkgs user which can't for some reason upgrade to the new version.

**Comment by willcohen** [comment]
That's a good question.

1) ofborg is breaking on at least linux too, so this PR Is not ready for merge. As they dropped autoconf the cmake configurations changed a little. May need a combination of some dylib fixes for darwin and some `$out` finagling overall -- still diagnosing.
2) Once I get point 1 resolved I'll see what `nixpkgs-review` yields. If stuff depending on this breaks I'll look into seeing what needs to be proj_8 and whether it's droppable.

**Comment by risicle** [comment]
FWIW many _many_ tests fail for proj 9 on nixos x86_64 for me. Many don't seem to be successfully loading the library at all: `/build/source/build/bin/test_tinshift: error while loading shared libraries: libproj.so.25: cannot open shared object file: No such file or directory`

**Comment by willcohen** [comment]
Yeah -- if I turn off `doCheck` on darwin, the `bin` files do seem to work correctly based on a few manual tests, and running `otool` on the generated proj dylib seems like it's pointing everything to the right place.

This leads me to think the problem is actually only about the testing, not the building.

**Comment by willcohen** [comment]
With the latest commits, the test suite now passes on `darwin` and `linux`. From here, working on seeing what packages are still dependent on `proj_8`, if any.

**Comment by willcohen** [comment]
On darwin, `nixpkgs-review` breaks with the following error:

```
error: attribute 'x86_64-darwin' missing

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/pkgs/tools/audio/essentia-extractor/default.nix:15:10:

           14|
           15|   arch = arch_table.${stdenv.system};
             |          ^
           16|   sha = sha_table.${stdenv.system};

       … while evaluating the attribute 'name'

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/pkgs/build-support/fetchurl/default.nix:125:3:

          124| stdenvNoCC.mkDerivation {
          125|   name =
             |   ^
          126|     if showURLs then "urls"

       … while evaluating the derivation attribute 'name'

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let

       … while evaluating the attribute 'src' of the derivation 'essentia-extractor-2.1_beta2'

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let

       … while evaluating the attribute 'postPatch' of the derivation 'acousticbrainz-client-0.1'

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let

       … while evaluating the attribute 'out.outPath'

       at /Users/wcohen/.cache/nixpkgs-review/pr-164879-1/nixpkgs/lib/customisation.nix:157:13:

          156|             drvPath = assert condition; drv.${outputName}.drvPath;
          157|             outPath = assert condition; drv.${outputName}.outPath;
             |             ^
          158|           };

       … while querying the derivation named 'acousticbrainz-client-0.1'
https://github.com/NixOS/nixpkgs/pull/164879 failed to build
```

Still trying to work on it with a linux machine instead, but it's pulling extremely slowly from cache so it looks like there's a lot of pre-proj dependencies that will still have to rebuild there for me.

**Comment by risicle** [comment]
`nixpkgs-review` reveals no new failures on macos 10.15. Still building for nixos.

**Comment by SuperSandro2000** [comment]
> On darwin, `nixpkgs-review` breaks with the following error:

Let me fix that.

File: pkgs/development/libraries/proj/default.nix [hunk 1]
```diff
@@ -36,6 +36,21 @@ stdenv.mkDerivation rec {
   preCheck = ''
     export HOME=$TMPDIR
     export TMP=$TMPDIR
+    # fix rpath for tests, needed since PROJ 9.0
+    for file in ./bin/*
+      do
+  '' + lib.optionalString stdenv.isDarwin ''
+      if ! [[ "$file" = ./bin/invgeod ]] &&
+         ! [[ "$file" = ./bin/proj ]]
+        then
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:45]:
```suggestion
      if [[ "$file" != ./bin/invgeod && "$file" != ./bin/proj ]]; then
```

File: pkgs/development/libraries/proj/default.nix [hunk 2]
```diff
@@ -36,6 +36,21 @@ stdenv.mkDerivation rec {
...
+          install_name_tool -add_rpath $(pwd)/lib $file
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:46]:
```suggestion
      if [[ "$file" != ./bin/invgeod && "$file" != ./bin/proj ]]; then
        install_name_tool -add_rpath $(pwd)/lib $file
```

File: pkgs/development/libraries/proj/default.nix [hunk 3]
```diff
@@ -36,6 +36,21 @@ stdenv.mkDerivation rec {
...
+  '' + lib.optionalString stdenv.isLinux ''
+       if ! [[ "$file" = ./bin/invgeod ]]
+        then
+          patchelf --set-rpath "$(patchelf --print-rpath $file):$(pwd)/lib" $file
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:50]:
```suggestion
       if [[ "$file" != ./bin/invgeod ]]; then
         patchelf --set-rpath "$(patchelf --print-rpath $file):$(pwd)/lib" $file
```

File: pkgs/development/libraries/proj/default.nix [hunk 4]
```diff
@@ -36,6 +36,21 @@ stdenv.mkDerivation rec {
...
+  '' + ''
+        fi
+      done
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:53]:
```suggestion
      fi
    done
```

File: pkgs/development/libraries/proj/default.nix [hunk 5]
```diff
@@ -36,6 +36,21 @@ stdenv.mkDerivation rec {
   preCheck = ''
     export HOME=$TMPDIR
     export TMP=$TMPDIR
+    # fix rpath for tests, needed since PROJ 9.0
+    for file in ./bin/*
+      do
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:41]:
```suggestion
    for file in ./bin/*; do
```

File: pkgs/development/libraries/proj/8.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  version = "8.2.1";
+
+  src = fetchFromGitHub {
+    owner = "OSGeo";
+    repo = "PROJ";
+    rev = version;
+    sha256 = "sha256-tnaIqYKgYHY1Tg33jsKYn9QL8YUobgXKbQsodoCXNys=";
+  };
+
+  outputs = [ "out" "dev"];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/8.nix:23]:
```suggestion
  outputs = [ "out" "dev" ];
```

**Comment by willcohen** [comment]
Suggestions integrated with latest commit. Dropping `proj_8` for now, re-running `nixpkgs-review` on nixos to ensure that nothing still breaks.

**Comment by risicle** [comment]
I'm still getting the same failure as ofborg on darwin-aarch64 though.

**Comment by willcohen** [comment]
At first glance, it's totally unclear to me what could be causing that.

If you change https://github.com/NixOS/nixpkgs/pull/164879/files#diff-ea61b58bcec284bd8187f4720c507ae458ff8dd6d67ea3ddd08b9f73d16d3e94R42 to only exclude `invgeod` and not `proj`, does it work then? 

The only reason I suggest that is that the failing test leads with:
`Running /tmp/nix-build-proj-9.0.0.drv-0/source/test/cli/test83 using /tmp/nix-build-proj-9.0.0.drv-0/source/build/bin/proj:`. Perhaps only intel needs proj to be excluded, but M1 needs the rpath fixed there to not crash?

**Comment by willcohen** [comment]
@NixOS/darwin-maintainers do any of you have a hunch about what's failing in the test suite here? It's hard for me to diagnose this, since I'm on the intel side of things and it's a strangely elusive test failure.

This isn't a deeply urgent PR, I'm just trying to rustle all the QGIS dependencies up to newer versions as I keep working on fixing the darwin build.

**Review by dotlambda — COMMENTED** [review]
Please add `-DNLOHMANN_JSON_ORIGIN=external` to `cmakeFlags` and `nlohmann_json` to `buildInputs`.

**Comment by uri-canva** [comment]
It passes for me on `aarch64-darwin` assuming it's part of the regular attribute build:
```
uribaghin@Uris-MacBook-Pro> /nix/store/v818fw1qqqnym00r6adpz3bjmhdn7cra-proj-9.0.0/bin/proj --help                         ~/github/NixOS/nixpkgs
Rel. 9.0.0, March 1st, 2022
<proj>: 
invalid option: --
program abnormally terminated
[1] uribaghin@Uris-MacBook-Pro> nix-build -A proj.passthru.tests                                                           ~/github/NixOS/nixpkgs
error: attribute 'tests' in selection path 'proj.passthru.tests' not found
[1] uribaghin@Uris-MacBook-Pro> nix-build -A proj.tests                                                                    ~/github/NixOS/nixpkgs
error: attribute 'tests' in selection path 'proj.tests' not found
```

**Comment by uri-canva** [comment]
Disregard my previous message, my system is a mix of `aarch64-darwin` and `x86_64-darwin` running with rosetta, will get back when I can get it all on `aarch64-darwin`.

**Comment by uri-canva** [comment]
Even running the test on its own works:
```
uribaghin@Uris-MacBook-Pro> test/cli/testproj /nix/store/ska46hqdipr7c1m7i5l6jgb9pf4xc84d-proj-9.0.0/bin/proj             ~/github/OSGeo/PROJ
============================================
Running test/cli/testproj using /nix/store/ska46hqdipr7c1m7i5l6jgb9pf4xc84d-proj-9.0.0/bin/proj:
============================================
doing tests into file testproj_out, please wait
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
diff testproj_out with testproj_out.dist
TEST OK
test file testproj_out removed
```

**Comment by uri-canva** [comment]
Looking at https://github.com/OSGeo/PROJ/blob/c0e2a9a49e01a8167944092a825044a3c5f468b9/test/cli/CMakeLists.txt it looks like all the tests using `proj` are failing, and only those.

**Comment by thefloweringash** [comment]
Looks like `install_name_tool` on a symlink with the `codesign` wrapper is making things confusing. The `preCheck` rewriting is hitting the `invproj -> proj` symlink and rewriting the contents of `proj`, and then the wrapper is replacing `invproj` with a correctly signed copy. This breaks the symlink and leaves `proj` with a broken signature.

But that's a bit of an aside. IIUC the `preCheck` phase is rewriting the binaries to refer to the temporary build dir. It's easier and cleaner to configure the dynamic linker to have temporary additional search paths. This following works for me on aarch64-darwin (didn't test anything else).

```nix
{
  # ...
  preCheck =
    let
      libPathEnvVar = if stdenv.isDarwin then "DYLD_LIBRARY_PATH" else "LD_LIBRARY_PATH";
    in
      ''
        export HOME=$TMPDIR
        export TMP=$TMPDIR
        export ${libPathEnvVar}=$PWD/lib
      '';
}
```

**Comment by willcohen** [comment]
Thanks @thefloweringash -- will test on x86 and force push, assuming it works!

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 164879` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>gplates</li>
    <li>merkaartor</li>
    <li>openorienteering-mapper</li>
    <li>python310Packages.basemap</li>
    <li>python39Packages.basemap</li>
  </ul>
</details>
<details>
  <summary>30 packages failed to build:</summary>
  <ul>
    <li>gdal (python39Packages.gdal)</li>
    <li>gnudatalanguage</li>
    <li>grass</li>
    <li>libLAS</li>
    <li>mapcache</li>
    <li>mapproxy</li>
    <li>mapserver</li>
    <li>octavePackages.octproj</li>
    <li>postgresql11Packages.postgis</li>
    <li>postgresql12Packages.postgis</li>
    <li>postgresql13Packages.postgis</li>
    <li>postgresql14Packages.postgis</li>
    <li>python310Packages.cartopy</li>
    <li>python310Packages.geopandas</li>
    <li>python310Packages.opensfm</li>
    <li>python310Packages.osmnx</li>
    <li>python310Packages.owslib</li>
    <li>python310Packages.pyproj</li>
    <li>python310Packages.rasterio</li>
    <li>python39Packages.cartopy</li>
    <li>python39Packages.fiona</li>
    <li>python39Packages.geopandas</li>
    <li>python39Packages.opensfm</li>
    <li>python39Packages.osmnx</li>
    <li>python39Packages.owslib</li>
    <li>python39Packages.pyproj</li>
    <li>python39Packages.rasterio</li>
    <li>python39Packages.worldengine</li>
    <li>sumo</li>
    <li>t-rex</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>gdal_2</li>
    <li>libgeotiff</li>
    <li>libspatialite</li>
    <li>osm2pgsql</li>
    <li>proj</li>
    <li>python310Packages.fiona</li>
    <li>python310Packages.gdal</li>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.worldengine</li>
    <li>python39Packages.python-mapnik</li>
    <li>survex</li>
  </ul>
</details>

**Comment by willcohen** [comment]
Still running nixpkgs-review on nixos. Looks like at the very least gdal 2.x fails to build. Need to test a little further -- currently unclear if it's an issue with master or if it needs proj 8.

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 164879` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>cloudcompare</li>
    <li>gdal_2</li>
    <li>octavePackages.octproj</li>
    <li>python310Packages.basemap</li>
    <li>python310Packages.worldengine</li>
    <li>python39Packages.basemap</li>
    <li>python39Packages.worldengine</li>
  </ul>
</details>
<details>
  <summary>57 packages built:</summary>
  <ul>
    <li>gdal (python39Packages.gdal)</li>
    <li>gmt</li>
    <li>gnudatalanguage</li>
    <li>gplates</li>
    <li>grass</li>
    <li>libLAS</li>
    <li>libgeotiff</li>
    <li>libosmscout</li>
    <li>libspatialite</li>
    <li>mapcache</li>
    <li>mapproxy</li>
    <li>mapserver</li>
    <li>merkaartor</li>
    <li>openorienteering-mapper</li>
    <li>osm2pgsql</li>
    <li>osmscout-server</li>
    <li>paraview</li>
    <li>pdal</li>
    <li>postgresql11Packages.postgis</li>
    <li>postgresql12Packages.postgis</li>
    <li>postgresql13Packages.postgis</li>
    <li>postgresql14Packages.postgis</li>
    <li>proj</li>
    <li>python310Packages.cartopy</li>
    <li>python310Packages.fiona</li>
    <li>python310Packages.gdal</li>
    <li>python310Packages.geopandas</li>
    <li>python310Packages.opensfm</li>
    <li>python310Packages.osmnx</li>
    <li>python310Packages.owslib</li>
    <li>python310Packages.pygmt</li>
    <li>python310Packages.pyproj</li>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.rasterio</li>
    <li>python39Packages.cartopy</li>
    <li>python39Packages.fiona</li>
    <li>python39Packages.geopandas</li>
    <li>python39Packages.opensfm</li>
    <li>python39Packages.osmnx</li>
    <li>python39Packages.owslib</li>
    <li>python39Packages.pygmt</li>
    <li>python39Packages.pyproj</li>
    <li>python39Packages.python-mapnik</li>
    <li>python39Packages.rasterio</li>
    <li>pytrainer</li>
    <li>qgis</li>
    <li>qgis-ltr</li>
    <li>qmapshack</li>
    <li>readosm</li>
    <li>saga</li>
    <li>spatialite_tools</li>
    <li>sumo</li>
    <li>survex</li>
    <li>t-rex</li>
    <li>udig</li>
    <li>valhalla</li>
    <li>xastir</li>
  </ul>
</details>

**Comment by willcohen** [comment]
A number of various GIS packages depending on proj broke over the course of the last month, and have been broken since before this PR, per hydra -- it seems very likely due to gcc 10->11 given the timing. I've gone ahead and marked all of the ones that fail on Linux as broken.

**Comment by dotlambda** [comment]
> marked all of the ones that fail on Linux as broken

I saw you only marked them broken on Linux. Do you know if they still work on macOS?

**Comment by willcohen** [comment]
Based on hydra, they appear to still be okay on darwin (with the exception of octproj, which fails on both).

**Comment by dotlambda** [comment]
A downside to marking them broken on Linux only is that it's unlikely someone will ever check whether they still work on macOS in the future due to the lack of Darwin manpower.

**Comment by willcohen** [comment]
@dotlambda looks like the broken-on-linux-only turns out to be correct. all the remaining broken darwin packages were broken on hydra master, so i think it's safe to generally conclude that proj isn't the fault. that said, since they're all python i hesitate to mark anything else broken on darwin since i suspect it's a more underlying python issue in general.

**Comment by toonn** [comment]
Fwiw, that sounds very reasonable since GCC updates would affect Darwin a lot less. The Darwin stdenv is Clang based and I remember touching PROJ when I bumped LLVM.

**Review by dotlambda — APPROVED** [review]

**Review by ravenjoad — APPROVED** [review]
The broken Octave package is likely due to Octave being updated, rather than proj. Need to look into that myself.
LGTM!

File: pkgs/applications/misc/osmscout-server/default.nix [hunk 7]
```diff
@@ -14,44 +14,23 @@ let
...
     rev = version;
-    sha256 = "0rpsi6nyhcz6bv0jab4vixkxhjmn84xi0q2xz15a097hn46cklx9";
+    sha256 = "sha256-I14nQL0H2rMga+RDdAjykqmf7QIZveA6oGWu5PfZ89s=";
     fetchSubmodules = true;
   };
 
-  # Two patches required to work with valhalla 3.1
-  patches = [
-    # require C++14 to match latest Valhalla
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/osmscout-server/default.nix:30]: Please remove the unused fetchpatch from inputs.

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise LGTM

File: pkgs/applications/misc/osmscout-server/default.nix (hunk 7 above)
Comment (willcohen) [reply pkgs/applications/misc/osmscout-server/default.nix:30]: Good catch. Done with latest force-push!

**Comment by willcohen** [comment]
`fetchpatch` change made, thanks!

**Review by ravenjoad — APPROVED** [review]

**Comment by willcohen** [comment]
Thank you!

**Pull request merged**